### PR TITLE
chore: added capimate and automation script

### PR DESCRIPTION
**Pull request by clintonmedbery**
Closes D2IQ-93749

**What problem does this PR solve?**:
This adds CAPIMate to kommander applications and automatically updates it when pre release is ran.

**Which issue(s) does this PR fix?**:
https://d2iq.atlassian.net/browse/D2IQ-93749

**Special notes for your reviewer**:
To test this, run:
`cd hack/release`
`go run main.go pre-release --repo ~/Documents/d2iq/kommander-applications --version v2.4.0`

Note that the CAPIMate tag in the `cm.yaml` has been updated to 2.4.0

- [x] No License Change (or NA).

### CODE REVIEW

**Comment by clintonmedbery** [comment]
I don't mind writing tests, but I see no way to run the existing tests. Anyone have guidance?

Also should we now remove CAPIMate from the kommander repo?

**Comment by cbuto** [comment]
> I don't mind writing tests, but I see no way to run the existing tests. Anyone have guidance?
> 
> Also should we now remove CAPIMate from the kommander repo?

the existing tests can be run with `make go-test` or `cd hack/release && go test ./..`

they are also ran in CI i.e. https://github.com/mesosphere/kommander-applications/actions/runs/3291879628/jobs/5426581929

**Comment by cbuto** [comment]
> Also should we now remove CAPIMate from the kommander repo?

It seems like CAPIMate would stay in the Kommander repo but we just override some of the values here, so everything would stay the same in the Kommander repo (from what I understand)

File: hack/release/cmd/prerelease/prerelease.go [hunk 1]
```diff
@@ -5,7 +5,9 @@ import (
 	"log"
 	"os"
 
+
```
Comment (gracedo) [inline hack/release/cmd/prerelease/prerelease.go:8]:
```suggestion
```

File: hack/release/pkg/constants/constants.go [hunk 2]
```diff
@@ -3,7 +3,7 @@ package constants
 const (
 	KommanderAppPath     = "./services/kommander/"
 	KommanderAppMgmtPath = "./services/kommander-appmanagement/"
-
+  CAPIMateDefaultVersion = "v0.0.0-dev.0"
```
Comment (gracedo) [inline hack/release/pkg/constants/constants.go:6]: you should be able to use `gofmt` to fix the formatting in these files

File: hack/release/pkg/updatecapimate/updatecapimate.go [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+	}
+	newContents := strings.Replace(string(read), constants.CAPIMateDefaultVersion, chartVersion, -1)
+
+	err = os.WriteFile(helmReleaseFilePath, []byte(newContents), 0)
+	if err != nil {
+		return err
+	}
+
+  return nil
+}
```
Comment (gracedo) [inline hack/release/pkg/updatecapimate/updatecapimate.go:38]:
```suggestion
}

```

File: services/kommander/0.4.0/defaults/cm.yaml [hunk 4]
```diff
@@ -82,6 +82,28 @@ data:
...
+          memory: 512Mi
+        requests:
+          cpu: 100m
+          memory: 128Mi
+      securityContext:
+        runAsUser: 65532
+      service:
+        annotations: {}
+        port: 8443
+        type: ClusterIP
```
Comment (gracedo) [inline services/kommander/0.4.0/defaults/cm.yaml:106]:
```suggestion
    capimate:
      image:
        tag: v0.0.0-dev.0
```
we should just override the tag here, the rest we can stick with the default chart values

**Comment by gracedo** [comment]
> Also should we now remove CAPIMate from the kommander repo?

The only thing we should probably do is leave the `tag` field blank (`tag: `). Everything else needs to remain there as the default chart values https://github.com/mesosphere/kommander/blob/main/chart/kommander/values.yaml#L208

File: hack/release/pkg/chartversion/chartversion_test.go [hunk 5]
```diff
@@ -57,6 +57,8 @@ func TestUpdateChartVersionsSuccessfully(t *testing.T) {
 		// Get the diff between the HRs
 		changes, err := diff.Diff(branchHr, testHr)
 		assert.Nil(t, err)
+		assert.NotZero(t, len(changes))
```
Comment (mikolajb) [inline hack/release/pkg/chartversion/chartversion_test.go:60]: nit: `NotEmpty`

File: hack/release/pkg/updatecapimate/updatecapimate.go [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mesosphere/kommander-applications/hack/release/pkg/constants"
+)
+
+func UpdateCAPIMateVersion(kommanderApplicationsRepo, chartVersion string) error {
```
Comment (mikolajb) [inline hack/release/pkg/updatecapimate/updatecapimate.go:12]:
```suggestion
func UpdateCAPIMateVersion(kommanderApplicationsRepo, containerImageVersion string) error {
```

File: hack/release/pkg/updatecapimate/updatecapimate.go [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	pathToChart := filepath.Join(constants.KommanderAppPath, "*/*/cm.yaml")
```
Comment (mikolajb) [inline hack/release/pkg/updatecapimate/updatecapimate.go:13]:
```suggestion
	pathToCM := filepath.Join(constants.KommanderAppPath, "*/*/cm.yaml")
```

File: hack/release/pkg/updatecapimate/updatecapimate.go [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+	matches, err := filepath.Glob(filepath.Join(kommanderApplicationsRepo, pathToChart))
+	if err != nil {
+		return err
+	}
+	if len(matches) == 0 {
+		return fmt.Errorf("no matches found for path %s (verify the kommander-applications repo path is correct)", pathToChart)
+	}
+	if len(matches) > 1 {
+		return fmt.Errorf("found > 1 match for path %s (there should only be one match)", pathToChart)
+	}
```
Comment (mikolajb) [inline hack/release/pkg/updatecapimate/updatecapimate.go:23]: You can combine it with `len(matches) != 1` to have a fewer lines to cover with tests.

File: hack/release/pkg/updatecapimate/updatecapimate.go [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+	helmReleaseFilePath := matches[0]
+
+	read, err := os.ReadFile(helmReleaseFilePath)
```
Comment (mikolajb) [inline hack/release/pkg/updatecapimate/updatecapimate.go:26]: Change `helmReleaseFilePath` variable name or:

```suggestion
	read, err := os.ReadFile(matches[0])
```

File: hack/release/pkg/updatecapimate/updatecapimate_test.go [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+	"path/filepath"
+	"testing"
+
+	"github.com/mesosphere/kommander-applications/hack/release/pkg/constants"
+	cp "github.com/otiai10/copy"
+	"github.com/stretchr/testify/assert"
+)
+
+const rootDir = "../../../../"
+
```
Comment (mikolajb) [inline hack/release/pkg/updatecapimate/updatecapimate_test.go:14]: To increase the test coverate
```suggestion

func TestUpdateCAPIMateVersionsFailsWhenItCannotFindCM(t *testing.T) {
	tmpDir, err := os.MkdirTemp("", "prerelease")
	assert.Nil(t, err)
	defer os.RemoveAll(tmpDir)

	updateToVersion := "v1.0.0"
	err = UpdateCAPIMateVersion(tmpDir, updateToVersion)
	assert.ErrorContains(t, err, "verify the kommander-applications repo path")
}

```

Comment (clintonmedbery) [reply hack/release/pkg/updatecapimate/updatecapimate_test.go:14]: Thanks!

**Review by mikolajb — APPROVED** [review]
🚀

File: hack/release/cmd/prerelease/prerelease.go [hunk 11]
```diff
@@ -22,17 +23,22 @@ func init() { //nolint:gochecknoinits // Initializing cobra application.
...
 			if err != nil {
 				return err
 			}
-			fmt.Fprintf(cmd.OutOrStdout(), "Updated Kommander chart version to %s", chartVersion)
+
+			err = updatecapimate.UpdateCAPIMateVersion(kommanderApplicationsRepo, chartVersionString)
+			if err != nil {
+				return err
+			}
+			fmt.Fprintf(cmd.OutOrStdout(), "Updated Kommander chart version to %s", chartVersionString)
```
Comment (cbuto) [inline hack/release/cmd/prerelease/prerelease.go:41]: super tiny nit
```suggestion
			fmt.Fprintf(cmd.OutOrStdout(), "Updated CAPIMate version to %s", chartVersionString)
```

**Review by cbuto — APPROVED** [review]

**Review by gracedo — APPROVED** [review]

**Comment by gracedo** [comment]
@mesosphere-mergebot backport release-2.3

**Comment by gracedo** [comment]
@mesosphere-mergebot backport release-2.2

**Pull request merged**